backend/init/init.go

```diff
@@ -73,3 +74,36 @@ func Set(name string, f func() backend.Backend) {
 
 	backends[name] = f
 }
+
+// deprecatedBackendShim is used to wrap a backend and inject a deprecation
+// warning into the Validate method.
+type deprecatedBackendShim struct {
+	backend.Backend
+	Message string
+}
+
+// Validate the Backend then add the deprecation warning.
+func (b deprecatedBackendShim) Validate(c *terraform.ResourceConfig) ([]string, []error) {
+	warns, errs := b.Backend.Validate(c)
+	warns = append(warns, b.Message)
+	return warns, errs
+}
+
+// DeprecateBackend can be used to wrap a backend to retrun a deprecation
+// warning during validation.
+func deprecateBackend(b backend.Backend, message string) func() backend.Backend {
+	if _, ok := b.(backend.Enhanced); ok {
+		panic("cannot use DeprecateBackend on an Enhanced Backend")
+	}
+
+	if _, ok := b.(backend.Local); ok {
+		panic("cannot use DeprecateBackend on a Local Backend")
+	}
```

Comment:
Is this because the deprecation wrapper type assertions for these interfaces due to using the `Backend` interface directly? If so, i think a comment here could be helpful to clarify that, since at first reading this feels rather arbitrary and I can see a future maintainer ending up in here after getting one of these panics and wondering what it's all about.